python-package/xgboost/data.py

```diff
@@ -0,0 +1,624 @@
+        if len(data.indices) != len(data.data):
+            raise ValueError('length mismatch: {} vs {}'.format(
+                len(data.indices), len(data.data)))
+        self._warn_unused_missing(data)
+        handle = ctypes.c_void_p()
+        _check_call(_LIB.XGDMatrixCreateFromCSCEx(
+            c_array(ctypes.c_size_t, data.indptr),
+            c_array(ctypes.c_uint, data.indices),
+            c_array(ctypes.c_float, data.data),
+            ctypes.c_size_t(len(data.indptr)),
+            ctypes.c_size_t(len(data.data)),
+            ctypes.c_size_t(data.shape[0]),
+            ctypes.byref(handle)))
+        return handle, feature_names, feature_types
+
+
+__dmatrix_registry.register_handler(
+    'scipy.sparse.csc', 'csc_matrix', CSCHandler)
+
+
+class NumpyHandler(DataHandler):
+    '''Handler of `numpy.ndarray`.'''
+    def _maybe_np_slice(self, data, dtype):
+        '''Handle numpy slice.  This can be removed if we use __array_interface__.
+        '''
+        try:
+            if not data.flags.c_contiguous:
+                warnings.warn(
+                    "Use subset (sliced data) of np.ndarray is not recommended " +
+                    "because it will generate extra copies and increase " +
+                    "memory consumption")
+                data = np.array(data, copy=True, dtype=dtype)
+            else:
+                data = np.array(data, copy=False, dtype=dtype)
+        except AttributeError:
+            data = np.array(data, copy=False, dtype=dtype)
+        return data
+
+    def transform(self, data):
+        return self._maybe_np_slice(data, self.meta_type), None, None
+
+    def handle_input(self, data, feature_names, feature_types):
+        """Initialize data from a 2-D numpy matrix.
+
+        If ``mat`` does not have ``order='C'`` (aka row-major) or is
+        not contiguous, a temporary copy will be made.
+
+        If ``mat`` does not have ``dtype=numpy.float32``, a temporary copy will
+        be made.
+
+        So there could be as many as two temporary data copies; be mindful of
+        input layout and type if memory use is a concern.
+
+        """
+        if not isinstance(data, np.ndarray) and hasattr(data, '__array__'):
+            data = np.array(data, copy=False)
+        if len(data.shape) != 2:
+            raise ValueError('Expecting 2 dimensional numpy.ndarray, got: ',
+                             data.shape)
+        # flatten the array by rows and ensure it is float32.  we try to avoid
+        # data copies if possible (reshape returns a view when possible and we
+        # explicitly tell np.array to try and avoid copying)
+        flatten = np.array(data.reshape(data.size), copy=False,
+                           dtype=np.float32)
+        flatten = self._maybe_np_slice(flatten, np.float32)
+        self.check_complex(data)
+        handle = ctypes.c_void_p()
+        _check_call(_LIB.XGDMatrixCreateFromMat_omp(
```

Comment:
We could warn on copy by checking if the data pointer is the same. I think this could be quite useful for telling people their memory usage is unnecessarily high.

Reply:
Or just use array interface.